src/url_search_params.cpp

```diff
@@ -0,0 +1,59 @@
+#include "ada.h"
+#include "ada/url_search_params.h"
+
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace ada {
+
+void url_search_params::remove(const std::string_view key) {
+  params.erase(
+      std::remove_if(params.begin(), params.end(),
+                     [&key](auto &param) { return std::get<0>(param) == key; }),
+      params.end());
+}
+
+void url_search_params::remove(const std::string_view key,
+                               std::string_view value) {
+  params.erase(std::remove_if(params.begin(), params.end(),
+                              [&key, &value](auto &param) {
+                                return std::get<0>(param) == key &&
+                                       std::get<1>(param) == value;
+                              }),
+               params.end());
+}
+
+void url_search_params::sort() const noexcept {
+  // TODO: Implement this
+}
+
```

Comment:
At a glance, the set function does not match the specification.

You should first search for a matching key, if you find it, you replace its value with 'value' and return (ignoring duplicates).

Otherwise, you emplace back.

Reply:
This is not correct. `Set` updates all values of entries with key `key`. For example:

```
> let params = new URLSearchParams('key=value&key=value2')
undefined
> params.set('key', 'value3')
undefined
> params
URLSearchParams { 'key' => 'value3' }
```

Reply:
@anonrig Ok.

Spec:

If [this](https://webidl.spec.whatwg.org/#this)’s [list](https://url.spec.whatwg.org/#concept-urlsearchparams-list) [contains](https://infra.spec.whatwg.org/#list-contain) any [tuples](https://infra.spec.whatwg.org/#tuple) whose name is name, then set the value of the first such [tuple](https://infra.spec.whatwg.org/#tuple) to value and [remove](https://infra.spec.whatwg.org/#list-remove) the others. Otherwise, [append](https://infra.spec.whatwg.org/#list-append) (name, value) to [this](https://webidl.spec.whatwg.org/#this)’s [list](https://url.spec.whatwg.org/#concept-urlsearchparams-list).

What if I have key=value&joe=jill and I set key to value2. I expect to get key= value2&joe=jill but I think you will get joe=jill&key= value2.

Reply:
So I do insist. I think that this implementation is not standard compliant.

Reply:
There is a web-platform test that is particularly testing for this: https://github.com/web-platform-tests/wpt/blob/master/url/urlsearchparams-set.any.js#L7

Reply:
> What if I have key=value&joe=jill and I set key to value2. I expect to get key= value2&joe=jill but I think you will get joe=jill&key= value2.

Yes this is correct. This is a bug.

Reply:
Fixed it in a new commit.